elkscmd/man/man1/ecalc.1
.TH ecalc 1
.SH NAME
ecalc \- simple integer calculator for ELKS
.SH SYNOPSIS
.B ecalc
.SH DESCRIPTION
.I ecalc
is a simple terminal integer calculator for ELKS. It uses 16-bit signed integers(-32768 to 32767).
.SH OPERATORS
.TP 20
.B +
Addition
.TP 20
.B -
Subtraction
.TP 20
.B *
Multiplication
.TP 20
.B /
Division
.TP 20
.B e
Exponent
.TP 20
.B ~
Bitwise NOT
.TP 20
.B &
Bitwise AND
.TP 20
.B |
Bitwise OR
.TP 20
.B ^
Bitwise XOR
.TP 20
.B ()
Parentheses
.TP 20
.B ans
Previous result
.SH EXAMPLES
.TP 20
.B Example 1: 
3*10e3, should print 3000
.TP 20
.B Example 2:
ans/2|1, should print 1501
.SH AUTHORS
.\" If you have contributed to this program, please add your name here
Benjamin Helle

elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/misc_utils/ecalc.c
/*
 * ecalc - simple integer calculator for ELKS
 * 
 * Copyright (c) 2025 Benjamin Helle
 * SPDX-License-Identifier: GPL-2.0-or-later
 * 
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* version 1.2 */
/* notice: int is always signed 16 bit*/

/* define tokens*/
#define TOKEN_NULL  0
#define TOKEN_EOF   1
#define TOKEN_NUM   2
#define TOKEN_LPAR  3
#define TOKEN_RPAR  4
#define TOKEN_ANS   5
/* arithmetic operators */
#define TOKEN_ADD   10
#define TOKEN_SUB   11
#define TOKEN_MUL   12
#define TOKEN_DIV   13
#define TOKEN_EXP   14
/* logical operators */
#define TOKEN_AND   20
#define TOKEN_OR    21
#define TOKEN_XOR   22
#define TOKEN_NOT   23

char* src;
int token;
int num;
int result;

/* built in pow function to reduce size */
int ipow(int x, int y) {
  /* note: this could end up overflowing. to fix: use long */
  if (y < 0) return 0; /* can't do negative*/
  if (y == 0) return 1;
  int r = 1;
  while (y > 0) { /* keep multiplying in a loop while y > 0 */
    r *= x;
    y--;
  }
  return r; /* x^y */
}

void error(char *s) {
  fprintf(stderr, "error: %s\n", s);
  exit(1); /* exit on error */
}

void next_token(void) 
{
  while (*src == ' ') src++; /* skip whitespaces */

  if (*src == '\0') { /* end of input */
    token = TOKEN_EOF;
    return;
  }

  if (isdigit(*src)) { /* number */
    num = atoi(src);
    token = TOKEN_NUM;
    while (isdigit(*src)) src++; /* skip other digits */
    return;
  }

  if (strncmp(src, "ans", 3) == 0) { /* previous result */
    token = TOKEN_ANS;
    src += 3;
    return;
  }

  switch(*src) { /* operators and parentheses */
    case '+': token = TOKEN_ADD; break;
    case '-': token = TOKEN_SUB; break;
    case '*': token = TOKEN_MUL; break;
    case '/': token = TOKEN_DIV; break;
    case '(': token = TOKEN_LPAR; break;
    case ')': token = TOKEN_RPAR; break;
    case 'e': token = TOKEN_EXP; break;
    case '&': token = TOKEN_AND; break;
    case '|': token = TOKEN_OR; break;
    case '^': token = TOKEN_XOR; break;
    case '~': token = TOKEN_NOT; break;
    /* add more tokens here */
    default: token = TOKEN_NULL;
  }

  src++; 
}

/* forward declarations */
int parse_ans(void);
int parse_term(void);
int parse_factor(void);
int parse_expr(void);
int parse_exponent(void);
int parse_and(void);
int parse_or(void);
int parse_xor(void);
int parse_not(void);

/* bitwise operations */

/* this forms a predence tree, from lowest to highest precedence
 * OR, AND, XOR, ADD/SUB, MUL/DIV, EXP, UNARY(eg. NOT)
*/

int parse_and(void) {
  int val = parse_term();
  if (token == TOKEN_AND) {
    next_token();
    val &= parse_term();
  }
  return val;
}

int parse_xor(void) {
  int val = parse_and();
  if (token == TOKEN_XOR) {
    next_token();
    val ^= parse_and();
  }
  return val;
}

int parse_or(void) {
  int val = parse_xor();
  if (token == TOKEN_OR) {
    next_token();
    val |= parse_xor();
  }
  return val;
}


int parse_exponent(void) 
{
  int val = parse_factor();
  while (token == TOKEN_EXP) {
    next_token();
    int rs = parse_factor();
    val = ipow(val, rs);
  }
  return val;
}

int parse_factor(void)
{
  /* higher precedence */
  int val = 0;
  if (token == TOKEN_NUM) {
    val = num;
    next_token();
  } else if (token == TOKEN_LPAR) {
    next_token();
    val = parse_expr();
    if (token == TOKEN_RPAR) next_token();
    else error("expected ')'");
  } else if (token == TOKEN_ADD) {
    next_token();
    val = parse_factor();
  } else if (token == TOKEN_SUB) {
    next_token();
    val = -parse_factor();
  } else if (token == TOKEN_NOT) {
    next_token();
    val = ~parse_factor();
  } else if (token == TOKEN_ANS) {
    next_token();
    val = result;
  }
  else error("expected number or an expression."); /* error if nothing matches */
  return val;
}

int parse_term(void)
{
  int val = parse_exponent(); /* if not exponent, parse_exponent will return parse_factor */
  
  while (token == TOKEN_MUL || token == TOKEN_DIV) {
    int op = token;
    next_token();
    int rs = parse_exponent(); /* right side */
    if (op == TOKEN_MUL) val *= rs;
    else {
      if (rs == 0) error("division by zero"); /* check if zero to avoid dbz error */
      val /= rs;
    }
  }

  return val;
}

int parse_expr(void) 
{
  /* lowest precedence */
  int val = parse_or(); /* will check if logical operation, if not then return parse_term */
  while (token == TOKEN_ADD || token == TOKEN_SUB) {
    int op = token;
    next_token();
    int rs = parse_or();
    if (op == TOKEN_ADD) val += rs;
    else val -= rs;
  }

  return val;
}

int main(int argc, char **argv) 
{
  printf("ecalc 1.2\nSee the man page for help. Ctrl+C to exit\n");
  char line[0xFF];
  while (1) { /* main loop */
    printf("> ");
    if (fgets(line, sizeof(line), stdin) == NULL) break;
    
    src = line;
    next_token();

    result = parse_expr(); /* get result */

    printf("= %d\n", result); /* and print it */
  }

  return 0;
}
